source/common/stream_info/filter_state_impl.cc

```diff
@@ -10,15 +10,15 @@ void FilterStateImpl::setData(absl::string_view data_name, std::shared_ptr<Objec
                               FilterState::StreamSharing stream_sharing) {
   if (life_span > life_span_) {
     if (hasDataWithNameInternally(data_name)) {
-      throw EnvoyException(
+      IS_ENVOY_BUG(
```

Comment:
Should we change the signature of this method to never throw and instead return a "success" boolean that must be checked by the clients? There's a lot of internal code catching exceptions.

Reply:
It's already considered an API violation; I'm not sure that returning a value indicating that you violated the API makes for nice usage. It also wouldn't be clear what to return from `getDataMutableGeneric`: if it fails because it's read-only, returning a nullptr will likely result in a crash.

My reasoning was that this should catch incorrect usage in tests, but probably the best way to proceed in production would be to allow the illegal operation. It shouldn't leave anything in an invalid state, it just means there are conflicting uses of the same filter state name.

But I'm open to other opinions on the best way to proceed.

Reply:
I bumped into this when setting server name (SNI) metadata in the filter state of the downstream connection to enable the upstream connection use the same SNI that was used in the downstream connection. The issue is that downstream connection filter state has lifetime `Connection` by definition, and I was surprised that `setData()` call from the router filter would throw an exception in this case (when `auto_sni` is also configured). I naively thought that the Connection level value would have precedence, or at least Envoy would silently ignore the 2nd `setData()` call in this case.

Reply:
@kyessenov and @jrajahalme what should we do here?

If "overrides" are allowed, my intuition is that a more narrowly scoped (eg Request is more narrow than Connection) value would be visible. 

But it would be much simpler to just return without any change in the places where we used to throw; that is less of a behavior change, and given that this is defined as an API violation, that seems sufficient. The only inconsistency is what to do for `getDataMutable` on read-only data. Do we return a nullptr and hope the caller checks it so they don't crash, or return the pointer and let the caller overwrite the data (but with ENVOY_BUG so hopefully it will be detected)? 

I'd like to change this in the way that is least likely to crash; that's the goal of removing the exceptions. I'd be ok with inconsistent behavior, given that any of the above will trigger ENVOY_BUG and shouldn't be done.

What do you guys think?

Reply:
Yeah, I agree that we should never crash - especially since the interaction is non-trivial so it means the crash is likely to show up late during integration. This happened to us with two filters (TCP and HTTP) setting "peer metadata" filter state.

Re: overrides I also think that should be allowed. But we shouldn't be making the behavior change in the same PR.

In general, it's fair to treat filter state mutability as "best-effort". It's not meant to protect or isolate against evil filters (whether or not those exist). They should be useful to make less programmatic errors in filters, but violations should be viewed as "unsafe casts".  Applied to your example, I think we should allow const-cast and permit read only data returned as mutable.

Reply:
IMO returning without throwing would be good behavior, maybe doing debug logs in cases where setting metadata that already exists on the parent, for example, and logging at warning level if getting mutable for read-only metadata?

Reply:
The ENVOY_BUG will log (and increment a stat).